### PR TITLE
fix form serializing and demo

### PR DESCRIPTION
**Pull request by notwaldorf**
This PR does an unfortunate number of things:
- fixes the serializing logic in the form. Now we handle duplicate names correctly, as well as native checkboxes and radio-buttons, the bane of my existence. :skull: 
- prettifies the demo and adds a whole bunch of extra native elements. like datalist. 👑
- adds some docs!

PTAL @cdata @morethanreal

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -10,81 +10,110 @@
...
 <head>
+  <title>iron-form demo</title>
 
-  <meta charset="utf-8">
-  <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
   <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1, user-scalable=yes">
-
-  <title>iron-form demo</title>
+  <meta name="mobile-web-app-capable" content="yes">
+  <meta name="apple-mobile-web-app-capable" content="yes">
```
Comment (cdata) [inline demo/index.html]: Should we be adding these to our other demos?

File: iron-form.html [hunk 2]
```diff
@@ -11,12 +11,35 @@
...
+call the form's `submit` method.
+
+  Example:
+
+    <paper-button raised onclick="submitForm()">Submit</paper-button>
+
+    function submitForm() {
+      document.getElementById('form').submit();
+    }
+*/
```
Comment (cdata) [inline iron-form.html]: Didn't we fix the issue that forces us to use JavaScript comments for docs?

File: iron-form.html [hunk 3]
```diff
@@ -88,25 +111,45 @@
 
     /**
      * Returns a json object containing name/value pairs for all the registered
-     * custom components and native elements of the form.
+     * custom components and native elements of the form. If there are elements
+     * with duplicate names, then their values will get aggregated into an
+     * array of values.
```
Comment (cdata) [inline iron-form.html]: Nice :goat:

File: iron-form.html [hunk 4]
```diff
@@ -88,25 +111,45 @@
...
     serialize: function() {
       var json = {};
 
+      function addSerializedElement(el) {
+        // If the name doesn't exist, add it. Otherwise, serialize it to
+        // an array,
+        if (!json[el.name]) {
+          json[el.name] = el.value;
+        } else {
+          if (typeof(json[el.name]) !== 'object') {
```
Comment (cdata) [inline iron-form.html]: Maybe use `!Array.isArray(json[el.name])`?

File: demo/index.html (hunk 1 above)
Comment (notwaldorf) [inline demo/index.html]: Oh, i copied them from the `paper-checkbox demo`. I have no idea. :(

File: iron-form.html (hunk 4 above)
Comment (notwaldorf) [inline iron-form.html]: Oh I didn't know there was a thing. Thank god, that `typeof` was giving me the creeps.

File: iron-form.html [hunk 5]
```diff
@@ -130,6 +173,15 @@
         valid = el.validate() && valid;
       }
       return valid;
+    },
+
+    _useValue: function(el) {
+      // Checkboxes and radio buttons should only use their value if they're checked.
+      if (el.type !== 'checkbox' && el.type !== 'radio') {
```
Comment (cdata) [inline iron-form.html]: This almost seems like something that could be implemented by the input. For example, the input behavior could have a `hasUsableValue` method that can always return true by default, but can be overridden by `checkbox` and `radio` types to use a different heuristic (e.g., `return this.checked`).

**Comment by notwaldorf** [comment]
Also, I just tried this locally, and you can't type into `textareas` at all in the demo :(

File: iron-form.html (hunk 5 above)
Comment (notwaldorf) [inline iron-form.html]: The problem is that the native checkbox can't override this. The only elements that currently pose this problem, really, are the native ones. (the `paper-checkbox` can just set a civilized true or false in `value`, for example)

File: iron-form.html (hunk 2 above)
Comment (notwaldorf) [inline iron-form.html]: I don't know if it's been tagged, since locally it doesn't seem to work

File: iron-form.html (hunk 5 above)
Comment (cdata) [inline iron-form.html]: Ahh, understood. I thought that all elements in the form were assumed to implement the appropriate behavior.

File: iron-form.html (hunk 2 above)
Comment (cdata) [inline iron-form.html]: Lame, okay.

**Comment by cdata** [comment]
LGTM :+1:

**Pull request merged**